### PR TITLE
test: add mmcs benchmark

### PR DESCRIPTION
**Pull request by batzor**
# Description

In this PR, we introduce a benchmark of our `TwoAdicFRI` implementation.

### CODE REVIEW

File: benchmark/fri/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Fri PCS Benchmark
```
Comment (TomTaehoonKim) [inline benchmark/fri/README.md:1]:
```suggestion
# FRI Benchmark
```

Comment (chokobole) [inline benchmark/fri/README.md:1]:
```suggestion
# FRI PCS Benchmark
```

File: benchmark/fri/fri_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#include "tachyon/crypto/hashes/sponge/truncated_permutation.h"
+#include "tachyon/math/finite_fields/baby_bear/baby_bear4.h"
+#include "tachyon/math/finite_fields/baby_bear/packed_baby_bear4.h"
+#include "tachyon/math/finite_fields/baby_bear/poseidon2.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/math/polynomials/univariate/radix2_evaluation_domain.h"
+
+namespace tachyon::benchmark {
+
+extern "C" tachyon_baby_bear* run_fri_pcs_plonky3_baby_bear(
```
Comment (TomTaehoonKim) [inline benchmark/fri/fri_benchmark.cc:30]:
```suggestion
extern "C" tachyon_baby_bear* run_fri_plonky3_baby_bear(
```

File: benchmark/fri/fri_runner.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        base::Map(degrees, [this](size_t degree) {
+          return this->pcs_.GetNaturalDomainForDegree(degree);
+        });
+
+    base::TimeTicks start = base::TimeTicks::Now();
+    typename PCS::Commitment commit;
+    typename PCS::ProverData prover_data;
+    CHECK(pcs_.Commit(inner_domains, inner_polys, &commit, &prover_data));
+    reporter_.AddTime(vendor, base::TimeTicks::Now() - start);
+    return static_cast<std::array<F, 8>>(commit)[1];
```
Comment (chokobole) [inline benchmark/fri/fri_runner.h:64]: Just a question! do we need this casting?

File: benchmark/fri/fri_runner.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  FRIRunner(SimpleReporter& reporter, const FRIConfig& config, PCS& pcs)
+      : reporter_(reporter), config_(config), pcs_(pcs) {}
+
+  F Run(Vendor vendor, const math::RowMajorMatrix<F>& input) {
+    size_t max_degree = static_cast<size_t>(input.rows());
+    std::vector<size_t> degrees;
+    for (size_t d = max_degree >> (config_.input_num() - 1); d <= max_degree;
+         d <<= 1) {
+      degrees.push_back(d);
+    }
```
Comment (chokobole) [inline benchmark/fri/fri_runner.h:44]: Please make private method for this and use this for both `Run()` and `RunExternal()`.

File: benchmark/fri/plonky3/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "plonky3_fri_benchmark"
+version = "0.0.1"
+authors = ["The Tachyon Authors <tachyon-discuss@kroma.network>"]
+edition = "2021"
+description = """
+Plonky3 Fri Benchmark
```
Comment (chokobole) [inline benchmark/fri/plonky3/Cargo.toml:7]:
```suggestion
Plonky3 FRI Benchmark
```

File: benchmark/fri/plonky3/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        num_queries: 10,
+        proof_of_work_bits: 8,
+        mmcs: challenge_mmcs,
+    };
+
+    let pcs = MyPcs::new(log_n, Dft {}, val_mmcs, fri_config);
+    (pcs, Challenger::new(perm.clone()))
+}
+
+fn do_test_fri_pcs<Val, Challenge, Challenger, P>(
```
Comment (TomTaehoonKim) [inline benchmark/fri/plonky3/src/lib.rs:86]:
```suggestion
fn do_test_fri<Val, Challenge, Challenger, P>(
```

File: benchmark/fri/plonky3/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    let start = Instant::now();
+    let (commit, _data) = pcs.commit(domains_and_polys);
+    unsafe {
+        duration.write(start.elapsed().as_micros() as u64);
+    }
+    Box::into_raw(Box::new(commit)) as *mut CppBabyBear
+}
+
+#[no_mangle]
+pub extern "C" fn run_fri_pcs_plonky3_baby_bear(
```
Comment (TomTaehoonKim) [inline benchmark/fri/plonky3/src/lib.rs:116]:
```suggestion
pub extern "C" fn run_fri_plonky3_baby_bear(
```

File: benchmark/fri/fri_benchmark.cc [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+  MyPackedCompressor packed_compressor(std::move(packed_sponge));
+  MMCS mmcs(hasher, packed_hasher, compressor, packed_compressor);
+
+  ChallengeMMCS challenge_mmcs(
+      ExtMMCS(std::move(hasher), std::move(packed_hasher),
+              std::move(compressor), std::move(packed_compressor)));
+
+  crypto::FRIConfig<ChallengeMMCS> fri_config{config.log_blowup(), 10, 8,
+                                              challenge_mmcs};
+  MyPcs pcs = MyPcs(std::move(mmcs), std::move(fri_config));
```
Comment (chokobole) [inline benchmark/fri/fri_benchmark.cc:96]:
```suggestion
  MyPCS pcs = MyPCS(std::move(mmcs), std::move(fri_config));
```

File: benchmark/fri/fri_benchmark.cc [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+void Run(const FRIConfig& config) {
+  constexpr size_t kRate = 8;
+  constexpr size_t kChunk = 8;
+  constexpr size_t kN = 2;
+
+  using F = math::BabyBear;
+  using ExtF = math::BabyBear4;
+  using PackedF = math::PackedBabyBear;
+  using ExtendedPackedF = math::PackedBabyBear4;
```
Comment (chokobole) [inline benchmark/fri/fri_benchmark.cc:51]:
```suggestion
  using ExtF = math::PackedBabyBear4;
```

File: benchmark/fri/fri_benchmark.cc [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  using MyCompressor = crypto::TruncatedPermutation<Poseidon2, kChunk, kN>;
+  using MyPackedCompressor =
+      crypto::TruncatedPermutation<PackedPoseidon2, kChunk, kN>;
+  using MMCS = crypto::FieldMerkleTreeMMCS<F, MyHasher, MyPackedHasher,
+                                           MyCompressor, MyPackedCompressor, 8>;
+  using ExtMMCS =
+      crypto::FieldMerkleTreeMMCS<ExtF, MyHasher, MyPackedHasher, MyCompressor,
+                                  MyPackedCompressor, 8>;
+  using ChallengeMMCS = crypto::ExtensionFieldMerkleTreeMMCS<ExtF, ExtMMCS>;
+  using Challenger = crypto::DuplexChallenger<Poseidon2, 16, 8>;
```
Comment (chokobole) [inline benchmark/fri/fri_benchmark.cc:69]:
```suggestion
  using Challenger = crypto::DuplexChallenger<Poseidon2, 16, kRate>;
```

File: benchmark/fri/fri_benchmark.cc [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  using MyPackedHasher =
+      crypto::PaddingFreeSponge<PackedPoseidon2, kRate, kChunk>;
+  using MyCompressor = crypto::TruncatedPermutation<Poseidon2, kChunk, kN>;
+  using MyPackedCompressor =
+      crypto::TruncatedPermutation<PackedPoseidon2, kChunk, kN>;
+  using MMCS = crypto::FieldMerkleTreeMMCS<F, MyHasher, MyPackedHasher,
+                                           MyCompressor, MyPackedCompressor, 8>;
+  using ExtMMCS =
+      crypto::FieldMerkleTreeMMCS<ExtF, MyHasher, MyPackedHasher, MyCompressor,
+                                  MyPackedCompressor, 8>;
```
Comment (chokobole) [inline benchmark/fri/fri_benchmark.cc:67]:
```suggestion
  using MMCS = crypto::FieldMerkleTreeMMCS<F, MyHasher, MyPackedHasher,
                                           MyCompressor, MyPackedCompressor, kChunk>;
  using ExtMMCS =
      crypto::FieldMerkleTreeMMCS<ExtF, MyHasher, MyPackedHasher, MyCompressor,
                                  MyPackedCompressor, kChunk>;
```

**Review by GideokKim — APPROVED** [review]
LGTM

File: benchmark/fri/fri_runner.h (hunk 3 above)
Comment (batzor) [reply benchmark/fri/fri_runner.h]: Compiler seemed to be complaining but it seems to work now ¯\_(ツ)_/¯. I removed it.

**Review by chokobole — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM (...the Mac benchmark 💀)

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**